// @tamagui/config/dist/cjs/config.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var config_exports = {};
__export(config_exports, {
  configWithoutAnimations: () => configWithoutAnimations
});
module.exports = __toCommonJS(config_exports);
var import_shorthands = require("@tamagui/shorthands"), import_themes = require("@tamagui/themes"), import_fonts = require("./fonts"), import_media = require("./media");
const configWithoutAnimations = {
  defaultFont: "body",
  shouldAddPrefersColorThemes: !0,
  themeClassNameOnRoot: !0,
  themes: import_themes.themes,
  media: import_media.media,
  shorthands: import_shorthands.shorthands,
  tokens: import_themes.tokens,
  fonts: import_fonts.fonts,
  mediaQueryDefaultActive: import_media.mediaQueryDefaultActive,
  selectionStyles: (theme) => theme.color5 ? {
    backgroundColor: theme.color5,
    color: theme.color11
  } : null
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  configWithoutAnimations
});
